File: kotlin-typescript/karakum/nameResolvers/resolveCustomTransformersAfterDeclarationsItemTypeArgumentName.js
import ts from "typescript";
import * as karakum from "karakum";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey)
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    const typeReference = getParent(node)
    if (!typeReference) return null
    if (!ts.isTypeReferenceNode(typeReference)) return null
    if (!ts.isIdentifier(typeReference.typeName)) return null
    if (typeReference.typeName.text !== "TransformerFactory") return null

    const union = getParent(typeReference)
    if (!union) return null
    if (!ts.isUnionTypeNode(union)) return null

    let arrayType = getParent(union)
    if (ts.isParenthesizedTypeNode(arrayType)) {
        arrayType = getParent(arrayType)
    }
    if (!arrayType) return null
    if (!ts.isArrayTypeNode(arrayType)) return null

    let property = getParent(arrayType)
    if (ts.isTypeOperatorNode(property)) {
        property = getParent(property)
    }
    if (!property) return null
    if (!ts.isPropertySignature(property)) return null
    if (!ts.isIdentifier(property.name)) return null

    const propertyName = property.name.text
    if (propertyName !== "afterDeclarations") return null

    const interfaceNode = getParent(property)
    if (!interfaceNode) return null
    if (!ts.isInterfaceDeclaration(interfaceNode)) return null

    const parentName = interfaceNode.name.text
    if (parentName !== "CustomTransformers") return null

    return `CustomTransformersAfterDeclarationsItemTypeArgument`
}

File: kotlin-typescript/karakum/src/jsMain/kotlin/typescript/karakum/nameResolvers/resolveCustomTransformersAfterDeclarationsItemTypeArgumentName.kt
package typescript.karakum.nameResolvers

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.*

fun resolveCustomTransformersAfterDeclarationsItemTypeArgumentName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val typeReference = ensureNotNull(typeScriptService.getParent(node))
    ensure(isTypeReferenceNode(typeReference))

    val typeName = typeReference.typeName
    ensure(isIdentifier(typeName))
    ensure(typeName.text == "TransformerFactory")

    val union = ensureNotNull(typeScriptService.getParent(typeReference))
    ensure(isUnionTypeNode(union))

    val arrayType = ensureNotNull(
        typeScriptService.getParent(union)
            ?.let {
                if (isParenthesizedTypeNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isArrayTypeNode(arrayType))

    val property = ensureNotNull(
        typeScriptService.getParent(arrayType)
            ?.let {
                if (isTypeOperatorNode(it)) {
                    typeScriptService.getParent(it)
                } else it
            }
    )
    ensure(isPropertySignature(property))

    val propertyNameNode = property.name
    ensure(isIdentifier(propertyNameNode))
    ensure(propertyNameNode.text == "afterDeclarations")

    val interfaceNode = ensureNotNull(typeScriptService.getParent(property))
    ensure(isInterfaceDeclaration(interfaceNode))
    ensure(interfaceNode.name.text == "CustomTransformers")

    "CustomTransformersAfterDeclarationsItemTypeArgument"
}
